net_connection/src/protocol_wrapper.rs

```diff
@@ -0,0 +1,433 @@
+//! This module provides a higher level interface to p2p / network messaging
+//! basically handles serialization / deserialization from / to the core
+//! protocol message types (NamedBinary and Json).
+
+use serde_json;
+
+use super::protocol::Protocol;
+
+/// High level p2p / network message
+#[derive(Debug, Clone, PartialEq)]
+pub enum ProtocolWrapper {
+    /// if we don't have a high-level repr, just store the base protocol
+    RawProtocol(Protocol),
+
+    /// [send] request the current state from the p2p module
+    RequestState,
+
+    /// [recv] p2p module is telling us the current state
+    State(StateData),
+
+    /// [send] request the default config from the p2p module
+    RequestDefaultConfig,
+
+    /// [recv] the default config from the p2p module
+    DefaultConfig(ConfigData),
+
+    /// [send] set the p2p config
+    SetConfig(ConfigData),
+
+    /// [send] connect to the specified multiaddr
+    Connect(ConnectData),
+
+    /// [recv] notification of a peer connected
+    PeerConnected(PeerConnectData),
+
+    /// [send] send a message to another node on the network
+    SendMessage(SendData),
+
+    /// [recv] recv the response back from a previous `SendMessage`
+    SendResult(SendResultData),
+
+    /// [recv] another node has sent us a message
+    HandleSend(HandleSendData),
+
+    /// [send] send our response to a previous `HandleSend`
+    HandleSendResult(SendResultData),
+}
+
+/// upgrade a Protocol reference to a ProtocolWrapper instance
+impl<'a> From<&'a Protocol> for ProtocolWrapper {
+    fn from(p: &'a Protocol) -> Self {
+        if let Protocol::Json(json) = p {
+            let json: serde_json::Value = serde_json::from_str(json.into()).unwrap();
+            let method = &json["method"];
+
+            if method == "requestState" {
+                return ProtocolWrapper::RequestState;
+            } else if method == "state" {
+                return ProtocolWrapper::State(StateData {
+                    state: match json["state"].as_str() {
```

Comment:
Since there is no early return in any of those `match` expressions, there is no need for one. You can just use the methods supplied by `Option<T>` to the code more simple.

Reply:
For example:
```rust
json["state"].as_str().unwrap_or("undefined").to_string()
```

Reply:
:+1: updated, but then ended up deleting this code...